docs/src/references/agent/operation-workflow.md

```diff
@@ -286,4 +288,163 @@ next = ["restarting", "successful_restart", "failed_restart"]
+on_error = { status = "fatal_state", reason = "fail to run the script"}  # possibly some `on_error` and `on_kill` handlers
+next = ["state-1", "state-2", "state-3"]                                 # the list of status accepted as next status
+```
+
+- If the script is successful and its output returns some `status` and `reason` fields, these are used for the next state.
+- If the script is successful but its output contains no `status` field, then `on_error` is used.
+- If the script cannot be executed, is killed or returns a non-zero code,
+  then the `status` and `reason` fields are determined by the `on_error` and `on_kill` handlers
+  (or if none where provided using the default `on_error` and `on_kill` handlers).
+- Except for the `status` and `reason` fields,
+  the fields of the JSON object extracted from the script output are injected into the command state message payload
+  (in the successful as well as the failed cases).
+
+### Using a script to trigger a restart
+
+A workflow state can be handled using a *background script*.
+When executed, as a detached process, by the __tedge_agent__ no response nor exit status is expected from the script.
+In any case the workflow will be moved to the given next state.
+And this move to the next state is even persisted *before* the script is executed.
+This can notably be used to restart the device or the agent.
+After the restart, the workflow will resume in the state specified by the workflow.
+
+```toml
+["agent-restart"]
+background_script = "sudo systemctl restart tedge-agent"
+on_exec = "waiting-for-restart"
+
+["waiting-for-restart"]
+script = "/some/script.sh checking restart"
+next = ["waiting-for-restart", "successful_restart", "failed_restart"]
+```
+
+Note that:
+- No `on_exit` nor `on_kill` status can be provided, as the script is not monitored.
+- If the script cannot be launched, the workflow will be moved to the final `"failed"` state.
+
+### Appropriate syntax for restart action
+
+First proposal:
+```toml
+["device-restart"]
+builtin_action = "restart"
+on_exec = "waiting-for-restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
+
+Pros:
+- closer to the internal behavior
+
+Cons:
+- state with no representation
+
+Alternative proposal:
+
+```toml
+["device-restart"]
+builtin_action = "restart"
+on_exec = "waiting-for-restart"
+
+["waiting-for-restart"]
+builtin_action = "waiting-for-restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
+
+Pros:
+- closer to the observed behavior
+- no implicit state
+
+Cons:
+- the two states must be consistent
+
+### Setting step execution timeout
+
+The execution time of the state transitions of a workflow can be limited using timeouts.
+- A default timeout can be set at the level of an operation for all the transitions.
+- Individual timeout can be set to each state of the workflow.
+
+```toml
+timeout_second = 300
+on_timeout = { status = "failed", reason = "timeout" }
```

Comment:
```suggestion
on_timeout = { duration_seconds = 300, status = "failed", reason = "timeout" }
```

How about merging them so that everything related to timeouts are at one place?